Fix #3406: [RTL] High-fi Recently- Played Activity List

veena14cs opened the pull request:
## Explanation
This PR fixes #3406 . This PR fixes rtl support in `ongoingStoryCard`.

## Screenshot LTR and RTL
Mobile:
<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354850-eb209700-e381-11eb-9097-d93d6b97de24.png">.......<img width="144" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354932-04c1de80-e382-11eb-8219-09981390189a.png">

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125355707-f6c08d80-e382-11eb-9784-72f776e4b5d3.png">....<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125356190-7f3f2e00-e383-11eb-8a14-90eaf3fcb111.png">



Tablet:

<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354465-76e5f380-e381-11eb-9363-f138dd8839cc.png">....<img width="160" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354177-2078b500-e381-11eb-8475-dc67faa57ca7.png">

<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354564-9250fe80-e381-11eb-934a-1fedcea6f38c.png">......<img width="256" alt="screenshot" src="https://user-images.githubusercontent.com/16301028/125354136-122a9900-e381-11eb-8c9d-326d896b1a4b.png">


## Tests with RTL code
<img width="384" alt="Screenshot (53)" src="https://user-images.githubusercontent.com/16301028/128327269-a8eca3e0-eb19-4a87-9837-83f2ca046dca.png">



## Tests without RTL code fails as expected
<img width="384" alt="Screenshot (54)" src="https://user-images.githubusercontent.com/16301028/128327203-14b904f0-fc18-4d80-bce4-464746e26644.png">



## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
fix rtl support in ongoing story card

veena14cs committed:
Update ongoing_story_card.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs committed:
Update OngoingStoryViewModel.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Update MarginBindingAdapters.java

veena14cs committed:
Update OngoingListAdapter.kt

rt4914 commented on app/src/main/res/layout-land/ongoing_story_card.xml:
```diff
@@ -59,8 +63,9 @@
           android:layout_marginEnd="8dp"
           android:ellipsize="end"
-          android:maxLines="1"
+          android:lines="2"
```
Can you confirm these in mocks? Share link of the mock where it has been shown as `2 lines`.

veena14cs replied:
In the link https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0e6ba726-1930-4f39-9c73-51df10a1c766/specs/ number of lines is not added in the mock. To keep it consistent with other versions of this xml files I changed it to `lines=2`

rt4914 replied:
Yes. It is 2 lines only.

rt4914 commented on app/src/main/res/layout-land/ongoing_story_card.xml:
```diff
@@ -59,8 +63,9 @@
-          android:maxLines="1"
+          android:lines="2"
           android:text="@{viewModel.ongoingStory.storyName}"
+          android:textStyle="bold"
```
Also for this change we need mock link

veena14cs replied:
Text is bold in the mock link in the link https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0e6ba726-1930-4f39-9c73-51df10a1c766/specs/ .

rt4914 replied:
Thanks.

rt4914 commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/OngoingListAdapter.kt:
```diff
@@ -61,104 +58,7 @@ class OngoingListAdapter(
       VIEW_TYPE_SECTION_STORY_ITEM -> {
         storyGridPosition = position - titleIndex
         (holder as OngoingStoryViewHolder).bind(itemList[position] as OngoingStoryViewModel)
-        val marginMin =
```
Thanks for getting rid of this really complex code. :-D

veena14cs replied:
:)

rt4914 reviewed (changes requested):
@veena14cs Great work. Suggest need mock links to verify the xml changes are correct or not.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

rt4914 reviewed (approved):
Nice work. Thanks

veena14cs commented:
> Nice work. Thanks

Thanks @rt4914 :)

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/OngoingListAdapter.kt:
```diff
@@ -61,104 +58,7 @@ class OngoingListAdapter(
-        holder.binding.ongoingStoryCardView.requestLayout()
-      }
-      else -> throw IllegalArgumentException("Invalid item view type: ${holder.itemViewType}")
+      }else -> throw IllegalArgumentException("Invalid item view type: ${holder.itemViewType}")
```
Nit: missing space between the brace & ``else``.

veena14cs replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/OngoingStoryViewModel.kt:
```diff
@@ -1,18 +1,100 @@
+   *  |       |_____|     |_____|     |_____|     |_____|       |
+   *  |                                                         |
+   */
+  fun computeStartMargin(): Int {
```
Can you write tests for this & the next function to make sure it computes the correct margin values for different positions?

veena14cs replied:
Added tests for recyclerview item as it uses this function for setting start and end margins.

BenHenning reviewed (changes requested):
Thanks @veena14cs! I think this great.

Would it be possible to add tests at the recently played activity level to verify that the different layouts presented in your screenshots are indeed laid out correctly in tests?

Edit: also, I think you're running into the same Bazel test failures here that are affecting your other PR.

veena14cs commented:
> Edit: also, I think you're running into the same Bazel test failures here that are affecting your other PR.

Yes, as this also includes same change in MrginBindingAdapters. If the #3452 is fixed I think it will not have Bazel test failures.

veena14cs committed:
Update OngoingListAdapter.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
added testcases

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

rt4914 commented:
Un-assigning as we have discussed the solution in meeting.

veena14cs committed:
Fixed testcases

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Update RecentlyPlayedFragmentTest.kt

veena14cs commented:
> Un-assigning as we have discussed the solution in meeting.

Thanks I have updated testcases.

veena14cs commented:
> Thanks @veena14cs! I think this great.
> 
> Would it be possible to add tests at the recently played activity level to verify that the different layouts presented in your screenshots are indeed laid out correctly in tests?
> 

@BenHenning Added tests PTAL.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt:
```diff
@@ -26,6 +29,7 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
 import androidx.test.rule.ActivityTestRule
 import com.google.common.truth.Truth.assertThat
 import dagger.Component
+import kotlinx.android.synthetic.main.ongoing_story_card.view.*
```
We shouldn't use ktx synthetic references as they're now deprecated: https://developer.android.com/topic/libraries/view-binding/migration (though we don't use Jetpack view binding since we use AndroidX databinding, instead).

veena14cs replied:
It was unused import. Removed it.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt:
```diff
@@ -362,6 +374,256 @@ class RecentlyPlayedFragmentTest {
+
+  @Config(qualifiers = "sw600dp-port")
+  @Test
+  fun testRecentlyPlayedTestActivity_recentlyPlayedItemInRtl_tabletPortrait_rtlMarginIsCorrect() {
```
Nit: let's order the tests consistently, e.g.:

- portrait
- landscape
- tabletPortrait
- tabletLandscape

veena14cs replied:
Done.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt:
```diff
@@ -933,6 +1195,11 @@ class RecentlyPlayedFragmentTest {
   }
 
+  fun setUpTestFragment(activity: RecentlyPlayedActivity) {
+    activity.supportFragmentManager.beginTransaction().add(testFragment, TEST_FRAGMENT_TAG)
+      .commitNow()
```
```suggestion
    activity.supportFragmentManager
      .beginTransaction()
      .add(testFragment, TEST_FRAGMENT_TAG)
      .commitNow()
```

Wrapping should start at the first ``.`` starting the chain of operations.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt:
```diff
@@ -981,3 +1248,4 @@ class RecentlyPlayedFragmentTest {
     override fun getApplicationInjector(): ApplicationInjector = component
   }
 }
+
```
Nit: remove extra newline (only one is needed at the end of the file).

veena14cs replied:
Done

BenHenning reviewed (changes requested):
Thanks @veena14cs. Left a couple of comments. Regarding the new tests & the shown failures, I only see 2 of the RTL tests failing in the screenshot. Shouldn't all of them fail if the new RTL code is reverted?

veena14cs committed:
Update app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

veena14cs commented:
> Thanks @veena14cs. Left a couple of comments. Regarding the new tests & the shown failures, I only see 2 of the RTL tests failing in the screenshot. Shouldn't all of them fail if the new RTL code is reverted?

Actually, the other 4 tests were added later and I had added 2 RTL tests earlier and its screenshot. I missed to update. I will do it now.

Edit: Updated screenshots with new tests.

veena14cs committed:
Update RecentlyPlayedFragmentTest.kt

veena14cs committed:
Merge branch 'fix-recently-played' of https://github.com/veena14cs/oppia-android into fix-recently-played

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs commented:
> Un-assigning as we have discussed the solution in meeting.

Updated tests. PTAL

BenHenning commented:
Thanks @veena14cs. I think this LGTM, but I can't confirm until #3452 is merged & this is updated to the latest develop. Please reassign to me at that point.

veena14cs commented:
> Thanks @veena14cs. I think this LGTM, but I can't confirm until #3452 is merged & this is updated to the latest develop. Please reassign to me at that point.

Thanks Ben.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

rt4914 reviewed (approved):
LGTM, thanks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

anandwana001 commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/OngoingStoryViewModel.kt:
```diff
@@ -1,18 +1,100 @@
 import android.view.View
+import androidx.appcompat.app.AppCompatActivity
 import androidx.lifecycle.ViewModel
+import org.oppia.android.R
```
with this addition, this file will go in the VIEW_MODEL_WTH_RESOURCES in app/BUILD.bazel file

veena14cs replied:
Done.

anandwana001 reviewed (changes requested):
Will review again, once the blocking RP get's in.

BenHenning commented:
Hi. Will need to look at this tomorrow.

Also, @veena14cs can you please fix the failing tests per CI?

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs commented:
I have merged with latest develop. This might fix CI.

veena14cs committed:
fixed nits

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
Update BUILD.bazel

BenHenning reviewed (approved):
Thanks @veena14cs. LGTM!

BenHenning commented:
(also assigning back to @anandwana001, not sure if you meant to do that earlier)

veena14cs commented:
> (also assigning back to @anandwana001, not sure if you meant to do that earlier)

Yes thanks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

anandwana001 reviewed (approved):
LGTM,
not the code owner for this PR, but keeping track of merging once all the checks pass.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into fix-recently-played

veena14cs committed:
Update RecentlyPlayedFragmentTest.kt
